src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,744 @@
+    }
+
+    // Determine the account we're sending from.
+    auto sendFromAccount = wallet.FindAccountForSelector(selector).value_or(ZCASH_LEGACY_ACCOUNT);
+
+    // We do not set a change address if there is no change.
+    std::optional<ChangeAddress> changeAddr;
+    auto changeAmount = spendable.Total() - resolvedPayments.Total() - fee;
+    if (changeAmount > 0) {
+        auto allowedChangeTypes = [&](const std::set<ReceiverType>& receiverTypes) -> std::set<OutputPool> {
+            std::set<OutputPool> result{resolvedPayments.GetRecipientPools()};
+            // We always allow shielded change when not sending from the legacy account.
+            if (sendFromAccount != ZCASH_LEGACY_ACCOUNT) {
+                result.insert(OutputPool::Sapling);
+            }
+            for (ReceiverType rtype : receiverTypes) {
+                switch (rtype) {
+                    case ReceiverType::P2PKH:
+                    case ReceiverType::P2SH:
+                        // TODO: This is the correct policy, but it’s a breaking change from
+                        //       previous behavior, so enable it separately.
+                        // if (strategy.AllowRevealedRecipients()) {
+                        if (!spendable.utxos.empty() || strategy.AllowRevealedRecipients()) {
+                            result.insert(OutputPool::Transparent);
+                        }
+                        break;
+                    case ReceiverType::Sapling:
+                        if (!spendable.saplingNoteEntries.empty() || strategy.AllowRevealedAmounts()) {
+                            result.insert(OutputPool::Sapling);
+                        }
+                        break;
+                    case ReceiverType::Orchard:
+                        if (params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+                            && (!spendable.orchardNoteMetadata.empty() || strategy.AllowRevealedAmounts())) {
+                            result.insert(OutputPool::Orchard);
+                        }
+                        break;
+                }
+            }
+            return result;
+        };
+
+        std::visit(match {
+            [&](const CKeyID& keyId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2PKH}));
+                if (sendTo.has_value()) {
+                    resolvedPayments.AddPayment(
+                            ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                    changeAddr = sendTo.value();
+                }
+            },
+            [&](const CScriptID& scriptId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2SH}));
```

Comment:
In thinking further about this, I don't think that `WalletTxBuilder` has the machinery to spend using arbitrary scripts to begin with, so this case should probably just return an error?

Reply:
Could it just use a fresh P2PKH address? That’s still transparent[^1].

What other issues would `WalletTxBuilder` have with sending from P2SH addresses? In that case, should we be preventing this earlier, rather than complaining about not being able to create a change address? (Because where would the failure occur if there were no change?)

[^1]: Is this maybe a good opportunity to opportunistically shield some change for users? Sure, the user may not have an account to shield to (or, we might not be able to guess which is the right one), but we could always shield to legacy Sapling, no? And that actually allows a stronger privacy policy (at least once #6409 is merged), since we already need `AllowRevealedSenders` for a P2SH spend (which subsumes `AllowRevealedAmounts`, needed for t->Sapling), but having transparent change would require `AllowFullyTransparent`.

Reply:
I suppose it’s not a surprise, but making `CScriptId` error here doesn’t cause any test failures.

Reply:
I just dug through this code a bit with @daira – we end up generating a fresh P2PKH addr for change for P2SH.

We also have test_bitcoin tests that use z_sendmany to spend P2SH.

Reply:
I can make that change, but I’m not sure it actually makes more sense. It’s not passing in where change should be sent to, but rather which “receivers” exist for the output. `allowedChangeTypes` is most useful for the NU5-style patterns (UAs, accounts), where we just pass in the set of receivers. For the legacy stuff, I think it makes sense to pass in the actual “receiver” and let the function figure out[^1] the right thing to do for that receiver type.

It might make more sense to rename `allowedChangeTypes` to something like `allowedChangePools`, which might also help clarify that it’s not a strict mapping where the change is limited to the receiver type (or even pool) of the `ZTXOPattern`).

Another question with `allowedChangeTypes` – I get why we want to allow shielded change whenever possible (but why Sapling instead of Orchard), but why do we initialize the set to allow any of the recipient pools as well?

[^1]: Currently all it does is map each receiver to an `OutputPool`, so `P2*` become `Transparent`.

Reply:
`allowedChangeTypes` makes sense because P2SH and P2PKH are in the same pool, but your explanation makes sense to me here; perhaps we should just add a comment to this effect.

We initialize the set to allow any of the recipient pools because the change output will be indistinguishable from the other outputs, but we'd prefer not to introduce a change output that is different from any of the existing recipient types (because that's going to reveal more information). 

In terms of the Sapling default, it's because we know that later we'll add Orchard if anything having to do with UAs is happening, but we always expect Sapling to be available. It might be that the empty set is a better default; this default was probably just from the need to have continuity before and after NU5.

Reply:
In any case, I'm now satisfied with this answer.

Reply:
> `allowedChangeTypes` makes sense because P2SH and P2PKH are in the same pool

I think this just might need to be verbified, because I think we have different interpretations of the currently ambiguous name. If we prefix with `add` (`addAllowedChangeTypes`), the name somewhat makes sense, because we’re providing the types we’re adding to the existing set of Sapling + recipients. But … 1. we’re not actually adding possible change types, we’re just indicating that these are the types of receiver our sending address has; 2. it’s not updating any state, which `add` would imply; and 3. this returns a value that is a set of pools, not a set of types. So, my suggestion is (implicitly until now) prefixing with `get`, in which case changing to `Pools` makes sense to do – `getAllowedChangePools`.

I agree it should have a comment as well, but I think that name change would help.